---
title: Quick start instructions for using core scripts to analyze CHIP-seq data
---

# Introduction
This document outlines how to use the scripts in the core_scripts/ directory to analyze data from a BMC experiment. This assumes the directory is clone which can be done using the git clone command. Furthermore, many of the scripts are meant to be run in a cluster environment (See README.md for assumptions). Thus, the user can use their institutions' cluster resources or install the softwares locally and adjust the scripts to use a for loop or a parallel library.
Each script carries out a particular operation along a chip-seq pipeline and requires a particular dependency.

# Protocol
1. Use setup scripts to create the metadata and documentation for the experiment. Synced the directory to the cluster.
    > 'setup_bmc_experiment.R', 'template_configuration_experiment_bmc.R', 'template_configuration_script_bmc.R'
    > Creates: '*_sample_grid.csv','*_bmc_table.tsv','*_configuration_experiment_bmc.R',

2. Download the data to the local directory according to the BMC's email instructions.
  > ngs_sync_data_from_bmc_to_personal_data_folder.sh
  - REQUIRES UPDATE (2025-11-10): Usually this would be all you need to do but if you submit two different orders and they get sequenced together, they will be delivered together. Use 'bmc_move_fastq_to_target_directory.sh' by modifying the variables and running as script.

3. Clean up the directory.
  > ngs_cleanup_experiment_directory.sh: Leaves only fastq files in the fastq subdirectory.

4. Verify fastq files for paired ends. Also checks the metadata extracted from the filename of the fastq files.
  > ngs_verify_fastq_and_output_manifest.sh

5. Consolidate the fastq files from lanes and rename the fastq files to consolidate_*.fastq convention.
  > Creates: 'consolidated*.fastq'

6. Submit script to go from fastq to bam and bigwig files.
  > Uses fastp, bowtie2, alignmentSieve, and bamcoverage.

7. Create plots from the bigwig files.
  > plot_genome_tracks_for_multiple_chromosomes_experiments_and_categories.R
  > Use the plot_*.R scripts to create the plots of the bigwig files as genomic tracks.

  > Scripts use 'configuration_experiment_bmc.R' and 'configuration_script_bmc.R'
  > Scripts use interactive repl. Source during R session. source('script.R')
  > View generated html, pdf or svg files (recommend pdf).
  > Creates: *.svg, *.html or *.pdf of plots generated via Gviz package.

10. Call peaks.
  > REQUIRES UPDATE (2025-11-10): Quick test of eaton peak calling replication.

11. TODO: Determine motifs.

12. TODO: Perform factor analysis.
